Threaded spec piecewise polynomial reader

geoffrey4444 opened the pull request:
## Proposed changes

Rewrites the ReadSpecThirdOrderPiecewisePolynomial action as a threaded action, so it reads the H5 file once / node, instead of once / element. Specifically, this PR introduces a new ThreadedAction that can be used similarly to the volume data reader's threaded action. Aside from the action, this scheme reuses everything (the VolumeDataReader parallel component and the SetData callback) already implemented for the VolumeDataReader.

The motivation is that this eliminates segfaults in the HDF5 library when reading in the spec function of time data in release mode.

### Types of changes:

- [x] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

geoffrey4444 committed:
Make spec FunctionOfTimeReader read once/node, not once/element

geoffrey4444 commented:
Some of the tests are giving me strange hdf5 errors...any advice how to fix?

wthrowe commented:
Make sure you're closing the file after writing it before calling the action.  Probably some braces around the writing part will fix it.

geoffrey4444 committed:
Use verbs charm on ocean

geoffrey4444 commented:
My apologies for taking so long to get back to this!

I just rebased and added braces around the file writing in the test (to force the file to close), but I still get "resource temporarily unavailable" in some tests. Locally, the tests pass, so I'm not sure how to fix this. Any other suggestions would be most welcome!

UPDATE: Thanks @nilsleiffischer for suggesting a fix for this problem, which I've now pushed, and that fixed the problem.

wthrowe commented on src/IO/Importers/ReadSpecThirdOrderPiecewisePolynomial.hpp:
```diff
@@ -22,15 +22,19 @@
 
 namespace importers {
-namespace Actions {
+namespace ThreadedActions {
```
This action isn't really threaded.  It forces all copies to run in serial.  Why are you making it a threaded action?

geoffrey4444 replied:
What I want to do is run the action once per node, then broadcast the data read to all processes. Previously, this action had no way to do this, but instead ran the action once / element. This led to hdf5 read errors in my BBH runs.

Since I didn't understand spectre's parallel infrastructure that well yet, I looked in the code for an example that read in data once/node, and the numeric initial data reader seemed the closest task for my purpose: read in data once/node one time at the beginning of the simulation. So I reused as much of the numeric initial data reader code as possible.

If there's a quick way to change this PR to get the same result with a better design, I'm all for it! But if it would take a lot of work to redesign, I'd rather do that as a followup PR

wthrowe replied:
The required change is changing "threaded" to "simple" where the action is called and removing the locking code and the extra scope.  (And changing the function arguments.  Probably just removing the lock, but I'm not sure.)

wthrowe commented on src/IO/Importers/ReadSpecThirdOrderPiecewisePolynomial.hpp:
```diff
@@ -59,147 +63,153 @@ namespace Actions {
+            spec_functions_of_time_for_tuple;
+        for (const auto& key_and_poly : spec_functions_of_time) {
+          spec_functions_of_time_for_tuple[key_and_poly.first] = static_cast<
+              std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>(
```
I don't think you need this cast.  (The conversion is not marked explicit.)

wthrowe commented on src/IO/Importers/ReadSpecThirdOrderPiecewisePolynomial.hpp:
```diff
@@ -59,147 +63,153 @@ namespace Actions {
+
+        // Package the domain::Tags::FunctionsOfTime in a TaggedTuple
+        tuples::tagged_tuple_from_typelist<
+            tmpl::list<::domain::Tags::FunctionsOfTime>>
```
`tuples::TaggedTuple<::domain::Tags::FunctionsOfTime>` (No need to get a typelist involved.)

wthrowe replied:
Why is this passed as a single-element tuple?  Why not just pass the data directly?

geoffrey4444 replied:
It's passed as a single-element tuple because the callback action that broadcasts the data to the elements, SetData, takes a tagged tuple (SetData is not always used to broadcast a single entity, as I'm using it here)

wthrowe reviewed (changes requested):
I think I need more explanation of how this is going to be used.  In particular, where is this action going to be called from?

geoffrey4444 commented:
> I think I need more explanation of how this is going to be used.  In particular, where is this action going to be called from?

It will be called from DgElementArray in a forthcoming BBH executable, just after calling the action to read numeric initial data. This code reuses the numeric initial data infrastructure and functions the same way

geoffrey4444 commented:
@wthrowe here is how I use this action currently in my BBH branch:


https://github.com/geoffrey4444/spectre/commit/1f15907209c9d66c8fc12d328a3bcc41b3f6e90b

It's completely analogous to the action that imports numeric initial data.

In the merged version, e.g. a compile time switch might control whether or not the action gets called, defaulting to not calling it

wthrowe commented:
As I've mentioned offline, I think the way the initial data is currently done is a bad design.  It could be done in a way similar to all the other actions, but instead we keep adding more and more special cases into DgElementArray.  I'd like to have a meeting about that design before we go further with this.

nilsvu commented:
@wthrowe we can talk after the code meeting tomorrow if that works for you?

kidder commented:
there is no meeting tomorrow as it is a holiday

geoffrey4444 commented:
I'm fine with revising the strategy for importing data, but could we perhaps go with the current design for this PR and then fix the design for both at once in a follow up PR? Currently, the non threaded version of the spec importer is just broken on multiple processors, and I'm very eager to get this fix in, because I'm eager to get BBH support (which needs this PR) on develop, not just on my branch.

wthrowe commented:
OK, but please make that a priority.  It's blocking the initialization changes I'm supposed to implement.

(And I don't know why you refer to this as threaded and the other as non-threaded. I think the old one was crashing because it was doing multi-threading and this fixes it by switching to a single thread.)

I'll finish my review in a bit.

geoffrey4444 commented:
> OK, but please make that a priority.  It's blocking the initialization changes I'm supposed to implement.
> 
> 
> 
> (And I don't know why you refer to this as threaded and the other as non-threaded. I think the old one was crashing because it was doing multi-threading and this fixes it by switching to a single thread.)
> 
> 
> 
> I'll finish my review in a bit.

Sounds good! Yes, I will make the redesign a priority! The stuff you're working on will help the bbh checkpointjng work.

Thanks for the review! What I meant to say was this version is aware of multi threading, accounts for oarallelization of our code,  the old one is not. Probably a better way than threaded to say that...

wthrowe commented on tests/Unit/IO/Importers/Test_ReadSpecThirdOrderPiecewisePolynomial.cpp:
```diff
@@ -219,76 +253,86 @@ SPECTRE_TEST_CASE("Unit.IO.ReadSpecThirdOrderPiecewisePolynomial",
+                       std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>
+        expected_functions_of_time;
+    expected_functions_of_time["ExpansionFactor"] =
+        static_cast<std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>(
```
I don't think this cast is necessary.

wthrowe commented on tests/Unit/IO/Importers/Test_ReadSpecThirdOrderPiecewisePolynomial.cpp:
```diff
@@ -219,76 +253,86 @@ SPECTRE_TEST_CASE("Unit.IO.ReadSpecThirdOrderPiecewisePolynomial",
+            std::make_unique<domain::FunctionsOfTime::PiecewisePolynomial<3>>(
+                expansion));
+    expected_functions_of_time["RotationAngle"] =
+        static_cast<std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>(
```
I don't think this cast is necessary.

wthrowe reviewed (changes requested):
Finishing the review I started above.

geoffrey4444 committed:
Fixup wthrowe review

geoffrey4444 commented:
After discussion with @nilsdeppe, I've completely rethought how the importing should work. I decided to close this PR and open a new one with the new approach (#2559) that supersedes this one.
